### PR TITLE
Fix colab and pybtex issue

### LINKED ISSUE #2524: Google Colab fails to install PyBaMM
### Discussed in https://github.com/pybamm-team/PyBaMM/discussions/2521
Google Colab has updated to Python 3.8 but now fails to install PyBaMM due to pybtex.

<div type='discussions-op-text'>

<sup>Originally posted by **reenubosepybamm** December  5, 2022</sup>
%pip install pybamm -q    # install PyBaMM if it is not installed
import pybamm
import numpy as np
import matplotlib.pyplot as plt
GIVING AN ERROR
PluginNotFound                            Traceback (most recent call last)
[<ipython-input-5-103384696aab>](https://localhost:8080/#) in <module>
      2 get_ipython().run_line_magic('pip', 'install pybamm --no-cache -q')
      3 get_ipython().run_line_magic('pip', 'install pybamm -q    # install PyBaMM if it is not installed')
----> 4 import pybamm
      5 import numpy as np
      6 import matplotlib.pyplot as plt

6 frames
[/usr/local/lib/python3.8/dist-packages/pybtex/plugin/__init__.py](https://localhost:8080/#) in _load_entry_point(group, name, use_aliases)
     79         for entry_point in pkg_resources.iter_entry_points(search_group, name):
     80             return entry_point.load()
---> 81     raise PluginNotFound(group, name)
     82 
     83 

PluginNotFound: plugin pybtex.database.input.bibtex not found</div>

### PR DESCRIPTION
**Pull request by Saransh-cpp**
# Description

This looks like this right now -

image

Now that I think of it, it might be better to not show a warning on importing pybamm (would be too noisy). Rather, we could do -

```py
    def __init__(self):
        # Set of citation keys that have been registered
        self._papers_to_cite = set()

        # Dict mapping citations keys to BibTex entries
        self._all_citations: dict[str, str] = dict()

        if "google.colab" not in sys.modules:
            self.read_citations()
            self._reset()
```
and then -
```py
def print_citations(filename=None, output_format="text"):
    """See :meth:`Citations.print`"""
    if "google.colab" in sys.modules:
        warnings.warn(
            """pybtex does not work with Google Colab due to a known bug -
            https://bitbucket.org/pybtex-devs/pybtex/issues/148/.
            Please manually cite all the references.""".replace(
                "\n                ", " "
            ),
        )
    else:
        pybamm.citations.print(filename, output_format)
```

Fixes #2524 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [ ] No style issues: `$ flake8`
- [ ] All tests pass: `$ python run-tests.py --unit`
- [ ] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: pybamm/citations.py [hunk 1]
```diff
@@ -143,7 +145,16 @@ def print(self, filename=None, output_format="text"):
...
 def print_citations(filename=None, output_format="text"):
     """See :meth:`Citations.print`"""
-    pybamm.citations.print(filename, output_format)
+    if "google.colab" in sys.modules:
+        raise ImportWarning(
+            """pybtex does not work with Google Colab due to a known bug -
+            https://bitbucket.org/pybtex-devs/pybtex/issues/148/.
+            Please manually cite all the references.""".replace(
+                "\n                ", " "
+            ),
```
Comment (tinosulzer) [inline pybamm/citations.py:154]:
```suggestion
            "pybtex does not work with Google Colab due to a known bug -"
            "https://bitbucket.org/pybtex-devs/pybtex/issues/148/."
            "Please manually cite all the references."
```

File: tests/unit/test_citations.py [hunk 2]
```diff
@@ -64,6 +66,12 @@ def test_print_citations(self):
         with self.assertRaisesRegex(pybamm.OptionError, "'text' or 'bibtex'"):
             pybamm.print_citations("test_citations.txt", "bad format")
 
+        # google colab issue - https://github.com/pybamm-team/PyBaMM/issues/2524
+        sys.modules["google.colab"] = "mock"
```
Comment (tinosulzer) [inline tests/unit/test_citations.py:70]: cool test!

**Review by tinosulzer — APPROVED** [review]
Looks good thanks! Will merge when tests pass

File: tests/unit/test_citations.py (hunk 2 above)
Comment (Saransh-cpp) [reply tests/unit/test_citations.py:70]: 😄

**Pull request merged**